File: indico/htdocs/js/indico/Core/Widgets/Navigation.js
/* This file is part of Indico.
 * Copyright (C) 2002 - 2014 European Organization for Nuclear Research (CERN).
 *
 * Indico is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * Indico is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Indico; if not, see <http://www.gnu.org/licenses/>.
 */

/**
 * Returns a list of page numbers
 * @param nPages The total number of pages through which we want to browse.
 * @param selectedPage The currently selected page.
 * @param around Number of pages that we want to whow to each side of the selected page. Note: if the selected page is 1,
 *               it will have around * 2 pages on its right.
 * @return A list with the page numbers that should appear on the page number widget.
 */
var makePageList = function(nPages, selectedPage, around) {
    var result = [];

    result.push(selectedPage);

    var min = selectedPage;
    var max = selectedPage;

    // create the near pages
    var remaining = around * 2;
    var i = 1;

    while (remaining > 0 && (min > 1 || max < nPages)) {
        if (selectedPage - i > 0) {
        	min = selectedPage - i;
            result.push(min);
            remaining--;
        }
        if (selectedPage + i <= nPages) {
        	max = selectedPage + i;
            result.push(max);
            remaining--;
        }
        i++;
    }

    // create the far pages
    var roundTo10 = function(n) {
        var remainder = n % 10;
        return (n - remainder);
    };

    min = roundTo10(min - 1);
    max = roundTo10(max + 10);

    var firstInserted = false;
    var lastInserted = false;

    for (i = 0; i < around/2; i++) {
    	if (min > 0) {
    		result.push(min)
    	}
    	if (max <= nPages) {
    		result.push(max)
    	}
    	min -= 10;
    	max += 10;
    }

    result.sort(numberSorter);

    if (result[result.length - 1] != nPages) {
    	result.push(nPages)
    }
    if (result[0] != 1) {
    	result.push(1);
    	result.sort(numberSorter);
    }

    return result;
};

File: indico/MaKaC/webinterface/navigation.py
# This file is part of Indico.
# Copyright (C) 2002 - 2016 European Organization for Nuclear Research (CERN).
#
# Indico is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 3 of the
# License, or (at your option) any later version.
#
# Indico is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Indico; if not, see <http://www.gnu.org/licenses/>.

import MaKaC.webinterface.urlHandlers as urlHandlers
from MaKaC import conference
from MaKaC.i18n import _

class NavigationEntry:
    _title = ""
    _url = None
    _parent = None

    @classmethod
    def getTitle(cls):
        return cls._title

    @classmethod
    def getURL(cls, *arg):
        if cls._url is not None:
            return cls._url.getURL(*arg)
        return cls._url

    @classmethod
    def getParent(cls, *arg):
        if cls._parent:
            return cls._parent()
        return None


class NEConferenceCFA( NavigationEntry ):
    _url = urlHandlers.UHConferenceCFA
    _title = "Call for Abstracts"

class NEAbstractSubmission( NavigationEntry ):
    _url = urlHandlers.UHAbstractSubmissionDisplay
    _parent = NEConferenceCFA
    _title = "Submit a new abstract"

class NEUserAbstracts( NavigationEntry ):
    _url = urlHandlers.UHUserAbstracts
    _parent = NEConferenceCFA
    _title = "View my abstracts"

class NEAbstractDisplay( NavigationEntry ):
    _url = urlHandlers.UHAbstractDisplay
    _parent = NEUserAbstracts
    _title = "Abstract details"

class NEContributionList( NavigationEntry ):
    _url = urlHandlers.UHContributionList
    _title = "Contribution List"


class NEMyStuff( NavigationEntry ):
    _url = urlHandlers.UHConfMyStuff
    _title = "My conference"

class NEAbstractModify( NavigationEntry ):
    _url = urlHandlers.UHAbstractModify
    _parent = NEAbstractDisplay
    _title = "Modification"

class NEAbstractWithdraw( NavigationEntry ):
    _url = urlHandlers.UHAbstractWithdraw
    _parent = NEAbstractDisplay
    _title = "Withdrawal"

class NEAbstractRecovery( NavigationEntry ):
    _url = urlHandlers.UHAbstractRecovery
    _parent = NEAbstractDisplay
    _title = "Recovery"


class NEAbstractSubmissionConfirmation( NavigationEntry ):
    _url = urlHandlers.UHAbstractSubmissionConfirmation
    _parent = NEConferenceCFA
    _title = "Abstract submission confirmation"
